### PR TITLE
USE_MKLDNN=1 is default in make build (mkldnn must be explicitly turned off)

### PR DESCRIPTION
**Pull request by azai91**
## Description ##
we are migrating to have mkldnn included in the default mxnet build. the USE_MKLDNN will be set to 1 by default (and therefore must explicitly be turned off if unsupported).

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] set USE_MKLDNN to 1 unless explicitly set to 0
- [ ] set USE_MKLDNN=0 for non-mkldnn builds in jenkins

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by szha** [comment]
@azai91 @mseth10 if possible please get results on more use cases. CIFAR for an image classification model along won't be representative enough.

**Comment by stu1130** [comment]
@mxnet-label-bot[pr-awaiting-response]

**Comment by pengzhao-intel** [comment]
I think @juliusshufan can help to provide the more accuracy and performance data.

**Comment by juliusshufan** [comment]
Update the  ImageNet-1k inference accuracy, based on Gluon-model zoo (pre-trained model), comparison target is NVidia-GPU. 
(Model including: AlexNet, VGG16, Resnet50-v1/v2, Inception-v3, DenseNet, SqueezeNet, MobileNetv1.0)

  | SKX | P40
-- | -- | --
  | python2 | python3 | python2 | python3
toplogy | top1 | top5 | top1 | top5 | top1 | top5 | top1 | top5
alexnet | 0.556455 | 0.785575 | 0.556455 | 0.785575 | 0.556455 | 0.785523 | 0.556455 | 0.785523
resnet50_v1 | 0.753367 | 0.926907 | 0.753367 | 0.926907 | 0.753367 | 0.926907 | 0.753367 | 0.926907
resnet50_v2 | 0.761327 | 0.929354 | 0.761327 | 0.929354 | 0.761327 | 0.929354 | 0.761327 | 0.929354
vgg16 | 0.720138 | 0.90662 | 0.720138 | 0.90662 | 0.720138 | 0.90662 | 0.720138 | 0.90662
densenet121 | 0.736587 | 0.917328 | 0.736587 | 0.917328 | 0.736587 | 0.917328 | 0.736587 | 0.917328
squeezenet1.1 | 0.561469 | 0.792099 | 0.561469 | 0.792099 | 0.561481 | 0.792099 | 0.561481 | 0.792099
mobilenet1.0 | 0.693531 | 0.889003 | 0.693531 | 0.889003 | 0.693531 | 0.889003 | 0.693531 | 0.889003
inceptionv3 | 0.762979 | 0.928074 | 0.762979 | 0.928074 | 0.762979 | 0.92814 | 0.762979 | 0.92814

**Comment by juliusshufan** [comment]
ImageNet-1k training, we use the script under "example\image-classification", training executed on both GPU and CPU, with same hyper-parameters, without further tunings, training curses on GPU and CPU pretty aligned. The followings are corresponding to ResNet50 v1 and v2.
image
image

**Comment by juliusshufan** [comment]
Apart from ImageNet-1k traning test, training test also been executed on small dataset, that includes:

  | Training set | Validation set | Classes | Source
-- | -- | -- | -- | --
CiFAR-10 | 50,000 | 10,000 | 10 | Released by MxNET official   http://data.mxnet.io/data/cifar10/
CiFAR-100 | 50,000 | 10,000 | 100 | Released by MxNET official   http://data.mxnet.io/data/cifar100.zip
sampled ImageNet | 100,200 | 10,000 | 200 | Sampled from ImageNet-1k and converted following the structure   and classes retrieved from tinyImageNet   (https://www.kaggle.com/c/tiny-imagenet/)

**Comment by juliusshufan** [comment]
**Benchmark data**
The benchmark data contains data collected on Linux and Mac, and compared between build with and w.o. MKLDNN, as the computation on a build w.o. MKLDNN is too slow, only the performance data selected CNN models are listed, benchmarking script based on example\image-classfication\benchmark_score.py.

**On MacOS,** the default compilation configurations disabling the OPENMP, below tables listing the perf datas on build with MKLDNN(OPENMP enabled), and the build without MKLDNN.
**_(HW is iMac Pro with 8-core Xeon-W)_**

VGG16

batch   size | MKLDNN-enabled | w.o. MKLDNN | boost-up
-- | -- | -- | --
1 | 20.913986 | 7.821254 | 267.40%
16 | 24.273071 | 8.438211 | 287.66%
32 | 24.704907 | 8.480799 | 291.30%
64 | 24.94608 | 8.524874 | 292.63%
128 | 25.074148 | 8.53283 | 293.86%
256 | 25.2629 | 8.535707 | 295.97%

Inception-v3

batch   size | MKLDNN-enabled | w.o. MKLDNN | boost-up
-- | -- | -- | --
1 | 41.431404 | 10.323434 | 401.33%
16 | 54.312317 | 10.665803 | 509.22%
32 | 54.604119 | 10.621378 | 514.10%
64 | 54.39568 | 10.605843 | 512.88%
128 | 54.410785 | 10.62466 | 512.12%
256 | 54.614424 | 10.616772 | 514.42%

Inception-V4

batch   size | MKLDNN-enabled | w.o. MKLDNN | boost-up
-- | -- | -- | --
1 | 20.715221 | 5.655873 | 366.26%
16 | 26.249734 | 5.779357 | 454.20%
32 | 26.197659 | 5.761883 | 454.67%
64 | 26.16153 | 5.771389 | 453.30%
128 | 26.247461 | 5.778834 | 454.20%
256 | 26.313875 | 5.77839 | 455.38%

ResNet-50

batch   size | MKLDNN-enabled | w.o. MKLDNN | boost-up
-- | -- | -- | --
1 | 41.70109 | 19.246681 | 216.67%
16 | 43.132788 | 20.854712 | 206.83%
32 | 41.613291 | 20.570733 | 202.29%
64 | 38.13329 | 20.652445 | 184.64%
128 | 38.839577 | 20.685878 | 187.76%
256 | 38.853521 | 20.68953 | 187.79%

MobileNet

batch   size | MKLDNN-enabled | w.o. MKLDNN | boost-up
-- | -- | -- | --
1 | 200.91608 | 36.047475 | 557.37%
16 | 287.614019 | 37.224849 | 772.64%
32 | 277.838051 | 36.914548 | 752.65%
64 | 274.474078 | 36.939298 | 743.04%
128 | 273.622323 | 37.04172 | 738.69%
256 | 273.445636 | 36.947783 | 740.09%

**Comment by azai91** [comment]
Can we try metrics with MacOS on AVX2 ISA? We are seeing performance drop enabling MKLDNN.

**Comment by juliusshufan** [comment]
> Just verifying. The above table is for Mac?

@azai91 
1. Yes, for benchmarking data, today I also update the datas collected on CentOS 7.4, the data you were reviewing is collected on an iMAC Pro.
2. I'll collect the Mac perf data on a AVX2 processor later on, as the issue you mentioning "performance drop enabling MKLDNN.", I suspected this is caused by the OPENMP is NOT enabled by default, @xinyu-intel is working on this and can provide more inputs.

**Comment by xinyu-intel** [comment]
Hi @azai91 , you can try the below building method on mac:
```
brew install llvm
# .bash_profile
export LIBRARY_PATH=/usr/local/Cellar/llvm/6.0.1/lib/
# config.mk
CC=/usr/local/Cellar/llvm/6.0.1/bin/clang CXX=/usr/local/Cellar/llvm/6.0.1/bin/clang++
# mkldnn.mk L40 before cmake
CC=/usr/local/Cellar/llvm/6.0.1/bin/clang CXX=/usr/local/Cellar/llvm/6.0.1/bin/clang++
# makefile
ifeq ($(USE_OPENMP), 1)
#             ifneq ($(UNAME_S), Darwin)
                              CFLAGS += -fopenmp
#             endif
Endif
```

**Comment by juliusshufan** [comment]
**RNN related data, including both accuracy, and performance benchmarking.**
**Accuracy**
1. GNMT model implemented by gluon-nlp (scripts\nmt\train_gnmt.py), IWMT2015 dataset, en-vi translation. The decoder-encoder is a 2-layer LSTM, per the model implemenation, as gluon.rnncell used, the MKLDNN FC can be covered as it is gluon.rnncell is an unfused kernel, below figure is the ppl trends collected on both GPU and CPU, with same hyper-parameters, the two curves aligned very well.
image

**Comment by juliusshufan** [comment]
@azai91 @szha That's all the data I currently uploaded, may I know your comments anything else helping to support the set of USE_MKLDNN as default. Thanks.

**Comment by szha** [comment]
While the speed-up looks solid, I noticed the following:
1. A difference in top-1 inference accuracy in this comment for squeezenet https://github.com/apache/incubator-mxnet/pull/12591#issuecomment-423105682
2. Higher variance in the training accuracy compared to GPU, and the lack of validation accuracy in this comment https://github.com/apache/incubator-mxnet/pull/12591#issuecomment-423889618
3. A clear difference in accuracy in https://github.com/apache/incubator-mxnet/pull/12591#issuecomment-423890706
4. Lack of comparison between regular builds and mkl builds which is what we should establish instead.

I also have the following questions regarding the results:
1. What does "multi-node" mean in the second diagram in this comment? https://github.com/apache/incubator-mxnet/pull/12591#issuecomment-423889618
2. What would be the results for more common CPUs?

Overall, I think these evaluation doesn't yet cover the most important question for this PR: can we say with confidence that by switching to USE_MKLDNN by default, our library can achieve speed-up without losing accuracy, for different CPUs?

**Comment by szha** [comment]
Note that for larger datasets it's unlikely that people would use it for training, so inference results with pre-trained models would suffice for the purpose of comparing mkl builds with regular builds.

**Comment by pengzhao-intel** [comment]
Thanks for looking into our data and I agree that the inference results are more important. 
@juliusshufan will follow up your question.

**Comment by lupesko** [comment]
I'd love to see this one merged, and MXNet users benefitting from improved performance on CPU, but I agree with comments made earlier by @szha that we need clear comparison for speed and accuracy between non-MKLDNN and MKLDNN.

I also suggest we document these benchmarks and results on [MXNet CWiki](https://cwiki.apache.org/confluence/display/MXNET/Apache+MXNet+Home) instead of in this issue - will be easier to see a full and up-to-date status there. @xinyu-intel if it makes sense to you, can you please document it there?

**Comment by pengzhao-intel** [comment]
@lupesko It's a good idea to document the benchmark results in the website rather than github.
How about creating a separate page under doc in https://mxnet.incubator.apache.org/ ?
I think this is the major interface for MXNet users.

**Comment by azai91** [comment]
@juliusshufan can you provide benchmarks comparing mkldnn vs non?

**Comment by juliusshufan** [comment]
@azai91 sure, some CNN perf/benchmark data already on the previous comments, do you mean more model coverage? And I'll also update the same content to the on CWiki page.

**Comment by pengzhao-intel** [comment]
@azai91 I will sync with @juliusshufan in local. Will launch the benchmark during weekend :)

**Comment by azai91** [comment]
@pengzhao-intel thanks for the update. let me know when you're done. planning on taking vote Tuesday or Wednesday.

**Comment by pengzhao-intel** [comment]
Latest data updated on this wiki page:
https://cwiki.apache.org/confluence/display/MXNET/MXNet+with+Intel+MKL-DNN+-+Performance+Benchmarking

@azai91 could you rebase the code?

**Comment by azai91** [comment]
Without MKLDNN is it is SIGNIFICANTLY slower
```
INFO:root:batch size  1, dtype float32, images/sec: 1.980962
INFO:root:batch size 32, dtype float32, images/sec: 2.444706
```
Could not wait for completion with resnet 50.

**Comment by azai91** [comment]
Results with mobilenet
```
ubuntu@ip-172-31-5-67:~/incubator-mxnet$ MXNET_MKLDNN_ENABLED=1 python example/image-classification/benchmark_score.py
Assertion failure at kmp_runtime.cpp(6481): __kmp_team_pool == __null.
OMP: Error #13: Assertion failure at kmp_runtime.cpp(6481).
OMP: Hint: Please submit a bug report with this message, compile and run commands used, and machine configuration info including native compiler and operating system versions. Faster response will be obtained by including all program sources. For information on submitting this issue, please see https://bugs.llvm.org/.
Assertion failure at kmp_runtime.cpp(6481): __kmp_team_pool == __null.
OMP: Error #13: Assertion failure at kmp_runtime.cpp(6481).
OMP: Hint: Please submit a bug report with this message, compile and run commands used, and machine configuration info including native compiler and operating system versions. Faster response will be obtained by including all program sources. For information on submitting this issue, please see https://bugs.llvm.org/.
Assertion failure at kmp_runtime.cpp(6481): __kmp_team_pool == __null.
OMP: Error #13: Assertion failure at kmp_runtime.cpp(6481).
OMP: Hint: Please submit a bug report with this message, compile and run commands used, and machine configuration info including native compiler and operating system versions. Faster response will be obtained by including all program sources. For information on submitting this issue, please see https://bugs.llvm.org/.
Assertion failure at kmp_runtime.cpp(6481): __kmp_team_pool == __null.
OMP: Error #13: Assertion failure at kmp_runtime.cpp(6481).
OMP: Hint: Please submit a bug report with this message, compile and run commands used, and machine configuration info including native compiler and operating system versions. Faster response will be obtained by including all program sources. For information on submitting this issue, please see https://bugs.llvm.org/.
Assertion failure at kmp_runtime.cpp(6481): __kmp_team_pool == __null.
OMP: Error #13: Assertion failure at kmp_runtime.cpp(6481).
OMP: Hint: Please submit a bug report with this message, compile and run commands used, and machine configuration info including native compiler and operating system versions. Faster response will be obtained by including all program sources. For information on submitting this issue, please see https://bugs.llvm.org/.
Assertion failure at kmp_runtime.cpp(6481): __kmp_team_pool == __null.
OMP: Error #13: Assertion failure at kmp_runtime.cpp(6481).
OMP: Hint: Please submit a bug report with this message, compile and run commands used, and machine configuration info including native compiler and operating system versions. Faster response will be obtained by including all program sources. For information on submitting this issue, please see https://bugs.llvm.org/.
Assertion failure at kmp_runtime.cpp(6481): __kmp_team_pool == __null.
OMP: Error #13: Assertion failure at kmp_runtime.cpp(6481).
OMP: Hint: Please submit a bug report with this message, compile and run commands used, and machine configuration info including native compiler and operating system versions. Faster response will be obtained by including all program sources. For information on submitting this issue, please see https://bugs.llvm.org/.
Assertion failure at kmp_runtime.cpp(6481): __kmp_team_pool == __null.
OMP: Error #13: Assertion failure at kmp_runtime.cpp(6481).
OMP: Hint: Please submit a bug report with this message, compile and run commands used, and machine configuration info including native compiler and operating system versions. Faster response will be obtained by including all program sources. For information on submitting this issue, please see https://bugs.llvm.org/.
Assertion failure at kmp_runtime.cpp(6481): __kmp_team_pool == __null.
OMP: Error #13: Assertion failure at kmp_runtime.cpp(6481).
OMP: Hint: Please submit a bug report with this message, compile and run commands used, and machine configuration info including native compiler and operating system versions. Faster response will be obtained by including all program sources. For information on submitting this issue, please see https://bugs.llvm.org/.
Assertion failure at kmp_runtime.cpp(6481): __kmp_team_pool == __null.
OMP: Error #13: Assertion failure at kmp_runtime.cpp(6481).
OMP: Hint: Please submit a bug report with this message, compile and run commands used, and machine configuration info including native compiler and operating system versions. Faster response will be obtained by including all program sources. For information on submitting this issue, please see https://bugs.llvm.org/.
Assertion failure at kmp_runtime.cpp(6481): __kmp_team_pool == __null.
OMP: Error #13: Assertion failure at kmp_runtime.cpp(6481).
OMP: Hint: Please submit a bug report with this message, compile and run commands used, and machine configuration info including native compiler and operating system versions. Faster response will be obtained by including all program sources. For information on submitting this issue, please see https://bugs.llvm.org/.
INFO:root:It may take some time to run all models, set --network to run a specific one
INFO:root:run batchsize [1, 32, 64, 128, 256] by default, set --batch-size to run a specific one
INFO:root:network: mobilenet
INFO:root:device: cpu(0)
/home/ubuntu/incubator-mxnet/python/mxnet/module/base_module.py:68: UserWarning: Data provided by label_shapes don't match names specified by label_names ([] vs. ['softmax_label'])
  warnings.warn(msg)
[23:42:08] ../src/operator/nn/mkldnn/mkldnn_base.cc:74: Allocate 8192 bytes with malloc directly
[23:42:08] ../src/operator/nn/mkldnn/mkldnn_base.cc:74: Allocate 32768 bytes with malloc directly
INFO:root:batch size  1, dtype float32, images/sec: 31.501611
INFO:root:batch size 32, dtype float32, images/sec: 194.704657
INFO:root:batch size 64, dtype float32, images/sec: 247.321861
INFO:root:batch size 128, dtype float32, images/sec: 276.045449
INFO:root:batch size 256, dtype float32, images/sec: 257.687046
```

**Comment by xinyu-intel** [comment]
@azai91 which compiler are you using to build mxnet with mkldnn on  m5a.24xlarge?

**Comment by azai91** [comment]
```
ubuntu@ip-172-31-5-67:~/incubator-mxnet/build$ /usr/bin/c++ --version
c++ (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by vandanavk** [comment]
@mxnet-label-bot add [MKLDNN]

**Comment by roywei** [comment]
@azai91 Thanks for the contribution, could you trigger CI again?

**Comment by apeforest** [comment]
Do you also need to update osx.mk? Please make sure it's working the same on Mac OS

**Comment by sandeep-krishnamurthy** [comment]
@azai91 - Thanks a lot for this PR.
What are the next steps here?

**Comment by mseth10** [comment]
@azai91 we can close this PR now?

**Comment by pengzhao-intel** [comment]
@mseth10 Yes, I think so. 
Next step, we will co-work to 
-  update the document and install page
-  make MKLDNN with the static link
-  make MKLDNN as default in the nightly build

@lupesko @azai91 @TaoLv @xinyu-intel @ZhennanQin 
What's your opinion?

File: ci/docker/runtime_functions.sh [hunk 1]
```diff
@@ -669,7 +662,6 @@ build_ubuntu_gpu_cmake() {
         -DUSE_CUDA=1                            \
         -DUSE_CUDNN=1                           \
         -DUSE_MKLML_MKL=0                       \
-        -DUSE_MKLDNN=0                          \
```
Comment (szha) [inline ci/docker/runtime_functions.sh]: I don't think this is supposed to be removed.

**Comment by azai91** [comment]
closing this PR as this is a duplicate of https://github.com/apache/incubator-mxnet/pull/13681

**Pull request closed without merging**